[RVV] add qs8-dwconv support for risc-v

ken-unger opened the pull request:
- Added support for qs8-dwconv and qs8-qc8w-dwconv for RVV
- Generator script can support qu8 but I've left those kernels out given past comments on qu8 being deprecated.
- Tested on qemu and Spacemit K1.

fbarchard commented on src/qs8-dwconv/unipass-rvv.c.in:
```diff
@@ -0,0 +1,100 @@
+      vint32m${LMUL}_t vacc = __riscv_vle32_v_i32m${LMUL}(w, vl);
+      w = (const void*) ((uintptr_t) w + vlmax * sizeof(int32_t));
+
+      for (int k=0; k<${KERNEL_TILE}; k++) {
```
code style... should have space around binary operators.   check all operators have a space around them

for (int k = 0; k < ${KERNEL_TILE}; k++) {

typically we use size_t for most integers... I'd check if other kernels use size_t or int
personal nit - ++k preferred

for (size_t k = 0; k < ${KERNEL_TILE}; ++k) {

ken-unger replied:
sounds good, will do.

steven-johnson replied:
XNNPACK has a `.clang-format` file, would it suffice just to follow its suggestions?

fbarchard replied:
We dont have a clang format that follows XNNPack coding conventions.  The main difference is XNNPack has long lines, more like objective C.  And clang format / code standards dont cover our mix of python, asm and C.

fbarchard reviewed (changes requested):
overall ok.  can you quote a performance number compared to scalar?

would prefer to see variations for LMUL so we can test against a few variations, depending on the exact RISCV code and choose the best.

Although qu8 is being discouraged, it hasnt gone away.  
In general QD8 is a preferred format... but we dont have a dwconv for that.
5x5 seems to be rare... only on mobilenet v3.  3x3 is far more common.
On many cpus 5x5 would be an issue for register pressure.

ken-unger commented:
In terms of performance are you referring to the qs8-dwconv-bench results?   Below is a snippet comparing the mobilenet v1 subset on a K1 (vlen=256).  Around 10x faster for this subset.

Running ./qs8-dwconv-bench
Run on (8 X 1600 MHz CPU s)
CPU Caches:
  L1 Instruction 32 KiB (x8)
  L1 Data 32 KiB (x8)
  L2 Unified 512 KiB (x2)
Load Average: 2.00, 2.10, 2.07
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                         Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:112/W:112/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:32/real_time         18568354 ns     18559505 ns           38 OPS=389.121M/s bytes=43.2581M/s cpufreq=1.6G
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:112/W:112/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:64/real_time          9099572 ns      9091252 ns           76 OPS=397.016M/s bytes=110.374M/s cpufreq=1.6G
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:56/W:56/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:128/real_time          17784443 ns     17786151 ns           39 OPS=406.273M/s bytes=45.235M/s cpufreq=1.6G
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:56/W:56/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:128/real_time           4580208 ns      4581827 ns          153 OPS=394.379M/s bytes=109.913M/s cpufreq=1.6G
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:28/W:28/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:256/real_time           8914670 ns      8915580 ns           78 OPS=405.25M/s bytes=45.4011M/s cpufreq=1.6G
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:28/W:28/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:256/real_time           2236056 ns      2236370 ns          313 OPS=403.911M/s bytes=113.686M/s cpufreq=1.6G
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:14/W:14/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:512/real_time           4444486 ns      4444885 ns          158 OPS=406.422M/s bytes=46.6556M/s cpufreq=1.6G
qs8_dwconv_9p4c__scalar_lrintf/mobilenet_v1/H:14/W:14/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:512/real_time           1118630 ns      1118777 ns          625 OPS=403.694M/s bytes=118.087M/s cpufreq=1.6G

qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:112/W:112/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:32/real_time          3234607 ns      3238992 ns          216 OPS=2.23376G/s bytes=248.324M/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:112/W:112/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:64/real_time           799176 ns       795711 ns          876 OPS=4.5205G/s bytes=1.25673G/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:56/W:56/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:128/real_time           1479186 ns      1479508 ns          475 OPS=4.88468G/s bytes=543.867M/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:56/W:56/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:128/real_time            408674 ns       411487 ns         1715 OPS=4.42G/s bytes=1.23185G/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:28/W:28/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:256/real_time            743153 ns       743798 ns          940 OPS=4.86127G/s bytes=544.62M/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:28/W:28/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:256/real_time            199850 ns       199890 ns         3501 OPS=4.51924G/s bytes=1.272G/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:14/W:14/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:512/real_time            351208 ns       351501 ns         1991 OPS=5.14321G/s bytes=590.42M/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:14/W:14/KH:3/KW:3/PH:2/PW:2/S:2/D:1/G:512/real_time            106398 ns       106365 ns         6580 OPS=4.2443G/s bytes=1.24153G/s cpufreq=1.6G
qs8_dwconv_9p8vc__rvv/mobilenet_v1/H:7/W:7/KH:3/KW:3/PH:2/PW:2/S:1/D:1/G:1024/real_time             210250 ns       210241 ns         3330 OPS=4.29569G/s bytes=540.614M/s cpufreq=1.6G

ken-unger commented:
In general my experience has been that a higher LMUL is always preferred.  I wasn't completely sure about the philosophy in adding the non-production kernels, so committed just the production ones.  

I've also been running within tflite (benchmark_model) to see the net result.  For mobilenet V1, and on the same K1 platform, running single threaded, we were previously at 105ms and now at 10ms.  

 [Node type]                                              [count]         [avg ms]  
 Convolution (NHWC, QC8) DWConv               13          105.745        (old)
 Convolution (NHWC, QC8) DWConv               13           10.168         (new)

ken-unger commented:
@fbarchard I'll do the coding style fix and also add qu8 support to this PR since that touches mostly the same files.

ken-unger commented:
New commit added;
- fixed the coding style noted.
- added qu8-dwconv support.  So we now have the full suite of qs8-dwconv, qs8-qc8w-dwconv and qu8-dwconv supported for rvv.  
- removed qs8_dwconv_25p8vc__rvv from bench/qs8-dwconv.cc  as I was encountering a heap corruption with this test.  However adding a benchmark case for scalar 25p4c also exhibits the same apparent heap corruption.  I've filed https://github.com/google/XNNPACK/issues/7657 for this issue.

bhbruce commented on src/qu8-dwconv/gen/qu8-dwconv-9p8vc-minmax-fp32-rvv.c:
```diff
@@ -0,0 +1,126 @@
+        i[k] += vlmax;
+        vint32m8_t vk32 = __riscv_vreinterpret_v_u32m8_i32m8(__riscv_vzext_vf4_u32m8(vk, vl));
+        vint32m8_t vi32 = __riscv_vreinterpret_v_u32m8_i32m8(__riscv_vzext_vf4_u32m8(vi, vl));
+        vk32 = __riscv_vsub_vx_i32m8(vk32, vkernel_zero_point, vl);
```
You could try to change the computation order:
```
1. load 8-bit vk
2. vksubzero_16 = vwsub(vk, vkernel_zero_point).
3. Load 8-bit vi
4. Widen vi to 16-bit => vi16
5. wmacc(vacc, vi16, vksubzero_16)
```

ken-unger replied:
@bhbruce I didn't make this change, since vkernel_zero_point is a signed value?  So we need to widen (and cast) before subtraction as there is no mixed datatype with vwsub.  Unless I've misunderstood something.

bhbruce replied:
Yes, you are right. Thanks.

fbarchard replied:
iirc the kernel zero point is unsigned 0 to 255.  it is subtracted from the kernel (weights) to produce a signed value, which is technically int16  -255 to 255.

another way to implement zero points is a second multiply accumulate of input * kzeropoint.  then subtract it outside the loop.  This allows 2 unsigned int8 multiples, but works best if you keep a 2nd set of accumulators.  The good thing is you only need to do MR multiplies... it is not sensitive to NR.

ken-unger replied:
I like that idea @fbarchard.  Will give that a try next round.

bhbruce commented on src/qu8-dwconv/gen/qu8-dwconv-9p8vc-minmax-fp32-rvv.c:
```diff
@@ -0,0 +1,126 @@
+
+      for (size_t k = 0; k < 9; ++k) {
+        vuint8m2_t vi = __riscv_vle8_v_u8m2(i[k], vl);
+        vuint8m2_t vk = __riscv_vle8_v_u8m2(w, vl);
```
NIT:  __riscv_vle8_v_u8m2((const uint8_t*) w, vl);
To avoid future compiler warning, you could change it.

ken-unger replied:
Done

bhbruce commented on src/qs8-dwconv/gen/qs8-dwconv-9p8vc-minmax-fp32-rvv.c:
```diff
@@ -0,0 +1,122 @@
+        w = (const void*) ((uintptr_t) w + vlmax * sizeof(int8_t));
+
+        i[k] += vlmax;
+        vacc = __riscv_vwmacc_vv_i32m8(vacc, __riscv_vwcvt_x_x_v_i16m4(vi, vl), __riscv_vwcvt_x_x_v_i16m4(vk, vl), vl);
```
Now: 
1. vi widen to 16
2. vk widen to 16
3. vwmacc

Possible change:
1. vtmp16 = wmul(vi, vk)
2. vwadd(vacc, vtmp16)

ken-unger replied:
Done.  I didn't see a measurable performance increase, which I think is due to register dependencies which is similar in both cases.  However I kept this new version

bhbruce replied:
Thank you for experiments.

fbarchard commented on src/configs/dwconv-config.c:
```diff
@@ -1247,6 +1281,20 @@ static void init_qu8_dwconv_config(void) {
+      const struct xnn_hardware_config* hardware_config = xnn_init_hardware_config();
+      qs8_dwconv_config[0].minmax.unipass = (xnn_dwconv_unipass_ukernel_fn) xnn_qu8_dwconv_minmax_fp32_ukernel_9p8vc__rvv;
+      qs8_dwconv_config[0].init.qu8 = xnn_init_qu8_conv_minmax_fp32_scalar_params;
+      qs8_dwconv_config[0].channel_tile = 8* hardware_config->vlenb / sizeof(int32_t);
```
add space around *
8 * hardware_config->vlenb

ken-unger replied:
Done

fbarchard replied:
still has no space?
qs8_dwconv_config[0].channel_tile = 8* hardware_config->vlenb / sizeof(int32_t);
should be 8 space * space:
qs8_dwconv_config[0].channel_tile = 8 * hardware_config->vlenb / sizeof(int32_t);

fbarchard commented on src/qs8-dwconv/gen/qs8-dwconv-25p8vc-minmax-fp32-rvv.c:
```diff
@@ -0,0 +1,202 @@
+//   Template: src/qs8-dwconv/unipass-rvv.c.in
+//   Generator: tools/xngen
+//
+// Copyright 2024 Microchip
```
// Copyright 2025 Microchip

ken-unger replied:
Done

ken-unger commented:
@fbarchard and @bhbruce thank you for your review.  I believe I have address all comments except for one regarding qu8 order of operations which I don't think can be implemented as suggested.

ken-unger commented:
@fbarchard please approve so this can be merged.  Thank you.

fbarchard reviewed (approved):
prefer see m4 variant as well, so we can evaluate on various hardware in the future.
m8 will often cause register pressure and not be faster.  But this kernel seems okay

ken-unger commented:
@fbarchard would you know what I need to do to have this approved PR merged?  I'm not clear on next steps.  Thank you for the review and help. 

Ken

ken-unger commented:
@dsharlet would you be able to merge this PR?  Being outside of Google I am unable to.  Thank you.
